library/std/src/io/mod.rs

```diff
@@ -386,20 +385,19 @@ where
         }
 
         match r.read(&mut g.buf[g.len..]) {
-            Ok(0) => {
-                ret = Ok(g.len - start_len);
-                break;
+            Ok(0) => return Ok(g.len - start_len),
+            Ok(n) => {
+                // We can't let g.len overflow which would result in the vec shrinking when the function returns. In
+                // particular, that could break read_to_string if the shortened buffer doesn't end on a UTF-8 boundary.
+                // The minimal check would just be a checked_add, but this assert is a bit more precise and should be
+                // just about the same cost.
+                assert!(n <= g.buf.len() - g.len);
```

Comment:
Would it make sense to slice the slice that's been passed into the `read` call in the first place and then take its length? As in:

```rust
let buffer = &mut g.buf[g.len..];
match r.read(buffer) { 
    // ...
    Ok(n) => g.len += buffer[..n].len(),
}
```

or something along those lines? It has an added benefit of also detecting when returned `n` is greater than the length of the length of the buffer provided, and may result in slightly less code generated overall.

Reply:
This check already is looking for n being greater than the length of the buffer, but agreed that structuring it like that is a bit more clear. Will update later today.